### PR TITLE
feat(JOML): migrate

### PR DESCRIPTION
**Pull request by pollend**
requires: https://github.com/MovingBlocks/Terasology/pull/4356

### CODE REVIEW

File: src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java [hunk 1]
```diff
@@ -385,10 +386,10 @@ private EntityRef createMultiBlockEntity(EntityRef mainBlockEntity, Vector3i mai
...
-    private Region3i createAABB(Vector3i mainLocation, Collection<Vector3i> memberLocations) {
-        Region3i aabb = Region3i.createFromMinAndSize(mainLocation, new Vector3i(1, 1, 1));
+    private BlockRegion createRegion(Vector3i mainLocation, Collection<Vector3i> memberLocations) {
+        BlockRegion aabb = new BlockRegion(mainLocation);
         for (Vector3i memberLocation : memberLocations) {
-            aabb = aabb.expandToContain(memberLocation);
+            aabb = aabb.union(memberLocation);
         }
         return aabb;
     }
```
Comment (jdrueckert) [inline src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java]: @skaldarnar Correct me if I'm wrong but that sounds like a candidate for a stream and unioning?

File: src/main/java/org/terasology/multiBlock/Basic3DSizeFilter.java [hunk 2]
```diff
@@ -16,9 +16,9 @@
 package org.terasology.multiBlock;
 
 import com.google.common.base.Predicate;
-import org.terasology.math.geom.Vector3i;
+import org.joml.Vector3i;
 
-public class Basic3DSizeFilter implements Predicate<org.joml.Vector3i> {
+public class Basic3DSizeFilter implements Predicate<Vector3i> {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/Basic3DSizeFilter.java]: I think this should be
```suggestion
public class Basic3DSizeFilter implements Predicate<Vector3ic> {
```

File: src/main/java/org/terasology/multiBlock/Basic3DSizeFilter.java [hunk 3]
```diff
@@ -32,7 +32,7 @@ public Basic3DSizeFilter(int horizontal1Size, int horizontal2Size, int minHeight
     }
 
     @Override
-    public boolean apply(org.joml.Vector3i value) {
+    public boolean apply(Vector3i value) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/Basic3DSizeFilter.java]:
```suggestion
    public boolean apply(Vector3ic value) {
```

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java [hunk 4]
```diff
@@ -39,13 +39,16 @@
...
+    private BlockEntityRegistry blockEntityRegistry;
+
     public LayeredMultiBlockFormItemRecipe(Predicate<EntityRef> itemFilter, Predicate<Vector2i> sizeFilter,
                                            Predicate<ActivateEvent> activateEventFilter, String prefab, MultiBlockCallback<int[]> callback) {
         this.itemFilter = itemFilter;
         this.sizeFilter = sizeFilter;
         this.activateEventFilter = activateEventFilter;
         this.prefab = prefab;
         this.callback = callback;
+        this.blockEntityRegistry = CoreRegistry.get(BlockEntityRegistry.class);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java:51]: Is this the only way to retrieve the `BlockEntityRegistry` or could we use the `@In` annotation?

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java [hunk 5]
```diff
@@ -85,22 +88,21 @@ public boolean processActivation(ActivateEvent event) {
...
         LayerDefinition layerDefinition = layerDefinitions.get(layerIndex);
         Predicate<EntityRef> entityFilter = layerDefinition.entityFilter;
-        int minX = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.east()).x;
-        int maxX = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.west()).x;
-        int minZ = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.south()).z;
-        int maxZ = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.north()).z;
+        int minX = getLastMatchingInDirection( entityFilter, basePosition, Direction.RIGHT.asVector3i()).x;
+        int maxX = getLastMatchingInDirection( entityFilter, basePosition, Direction.LEFT.asVector3i()).x;
+        int minZ = getLastMatchingInDirection( entityFilter, basePosition, Direction.FORWARD.asVector3i()).z;
+        int maxZ = getLastMatchingInDirection( entityFilter, basePosition, Direction.BACKWARD.asVector3i()).z;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java]: `Vector3i.north` is defined as `(0, 0, 1)` while `Direction.BACKWARD` is `(0, 0, -1)`. Did you fix some bug here, or is this a mistake in the migration? 

```suggestion
        int minZ = getLastMatchingInDirection( entityFilter, basePosition, Direction.BACKWARD.asVector3i()).z;
        int maxZ = getLastMatchingInDirection( entityFilter, basePosition, Direction.FORWARD.asVector3i()).z;
```

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java [hunk 6]
```diff
@@ -190,10 +191,10 @@ private boolean processDetectionForLayer(ActivateEvent event, int layerIndex, Ve
         return true;
     }
 
-    private Vector3i getLastMatchingInDirection(BlockEntityRegistry blockEntityRegistry, Predicate<EntityRef> entityFilter, Vector3i location, Vector3i direction) {
-        Vector3i result = location;
+    private Vector3i getLastMatchingInDirection( Predicate<EntityRef> entityFilter, Vector3ic location, Vector3ic direction) {
+        Vector3i result = new Vector3i(location);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java:195]:
```suggestion
        Vector3i result = new Vector3i(location);
        Vector3i tmp = new Vector3i();
```

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java [hunk 7]
```diff
@@ -190,10 +191,10 @@ private boolean processDetectionForLayer(ActivateEvent event, int layerIndex, Ve
...
         while (true) {
-            Vector3i testedLocation = new Vector3i(result.x + direction.x, result.y + direction.y, result.z + direction.z);
+            Vector3i testedLocation = result.add(direction, new Vector3i());
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java]:
```suggestion
            result.add(direction, tmp);
```

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java [hunk 8]
```diff
@@ -190,10 +191,10 @@ private boolean processDetectionForLayer(ActivateEvent event, int layerIndex, Ve
...
             EntityRef blockEntityAt = blockEntityRegistry.getBlockEntityAt(testedLocation);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java:199]:
```suggestion
            EntityRef blockEntityAt = blockEntityRegistry.getBlockEntityAt(tmp);
```

Comment (skaldarnar) [reply src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java:199]: and further down replace 
```java
            result = testedLocation;
```
with
```java
            result.set(tmp);
```

File: src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java (hunk 6 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/multiBlock/recipe/LayeredMultiBlockFormItemRecipe.java:195]: Let's make better use of vector allocation here.

File: src/main/java/org/terasology/multiBlock2/component/MultiBlockMainComponent.java [hunk 9]
```diff
@@ -50,7 +50,7 @@ public MultiBlockMainComponent(List<Vector3i> multiBlockMembers, Region3i aabb,
         return Collections.unmodifiableList(multiBlockMembers);
     }
 
-    public Region3i getAabb() {
+    public BlockRegion getAabb() {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/component/MultiBlockMainComponent.java]: should this be modifiable, or should this return a `BlockRegionc`?
```suggestion
    public BlockRegionc getAabb() {
```

File: src/main/java/org/terasology/multiBlock2/recipe/LayeredMultiBlockRecipe.java [hunk 10]
```diff
@@ -65,23 +68,23 @@ public T detectFormingMultiBlock(Vector3i location) {
...
         LayerDefinition layerDefinition = layerDefinitions.get(layerIndex);
         Predicate<EntityRef> entityFilter = layerDefinition.entityFilter;
-        int minX = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.east()).x;
-        int maxX = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.west()).x;
-        int minZ = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.south()).z;
-        int maxZ = getLastMatchingInDirection(blockEntityRegistry, entityFilter, basePosition, Vector3i.north()).z;
+        int minX = getLastMatchingInDirection(entityFilter, basePosition, Direction.RIGHT.asVector3i()).x;
+        int maxX = getLastMatchingInDirection(entityFilter, basePosition, Direction.LEFT.asVector3i()).x;
+        int minZ = getLastMatchingInDirection(entityFilter, basePosition, Direction.DOWN.asVector3i()).z;
+        int maxZ = getLastMatchingInDirection(entityFilter, basePosition, Direction.UP.asVector3i()).z;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/recipe/LayeredMultiBlockRecipe.java]: `north` and `south` are not `UP`/`DOWN`.
```suggestion
        int minZ = getLastMatchingInDirection(entityFilter, basePosition, Direction.BACKWARD.asVector3i()).z;
        int maxZ = getLastMatchingInDirection(entityFilter, basePosition, Direction.FORWARD.asVector3i()).z;
```

File: src/main/java/org/terasology/multiBlock2/recipe/LayeredMultiBlockRecipe.java [hunk 11]
```diff
@@ -138,17 +140,17 @@ private T processDetectionForLayer(int layerIndex, Vector3i basePosition) {
...
-    protected abstract T createMultiBlockDefinition(Region3i multiBlockRegion, int[] layerHeights);
+    protected abstract T createMultiBlockDefinition(BlockRegionc multiBlockRegion, int[] layerHeights);
 
-    private Vector3i getLastMatchingInDirection(BlockEntityRegistry blockEntityRegistry, Predicate<EntityRef> entityFilter, Vector3i location, Vector3i direction) {
-        Vector3i result = location;
+    private Vector3i getLastMatchingInDirection(Predicate<EntityRef> entityFilter, Vector3ic location, Vector3ic direction) {
+        Vector3i result = new Vector3i(location);
         while (true) {
-            Vector3i testedLocation = new Vector3i(result.x + direction.x, result.y + direction.y, result.z + direction.z);
+            Vector3i testedLocation = result.add(direction, new Vector3i());
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/recipe/LayeredMultiBlockRecipe.java]: Apply the same pattern with a `tmp` vector here as commented further up.

File: src/main/java/org/terasology/multiBlock2/recipe/UniformBaseMultiBlockRecipe.java [hunk 12]
```diff
@@ -39,29 +43,29 @@ protected UniformBaseMultiBlockRecipe(BlockEntityRegistry blockEntityRegistry, P
...
-        int minY = getLastMatchingInDirection(target, location, Vector3i.down()).y;
-        int maxY = getLastMatchingInDirection(target, location, Vector3i.up()).y;
-        int minZ = getLastMatchingInDirection(target, location, Vector3i.south()).z;
-        int maxZ = getLastMatchingInDirection(target, location, Vector3i.north()).z;
+        int minX = getLastMatchingInDirection(target, location, Direction.RIGHT.asVector3i()).x;
+        int maxX = getLastMatchingInDirection(target, location, Direction.LEFT.asVector3i()).x;
+        int minY = getLastMatchingInDirection(target, location, Direction.DOWN.asVector3i()).y;
+        int maxY = getLastMatchingInDirection(target, location, Direction.UP.asVector3i()).y;
+        int minZ = getLastMatchingInDirection(target, location, Direction.FORWARD.asVector3i()).z;
+        int maxZ = getLastMatchingInDirection(target, location, Direction.BACKWARD.asVector3i()).z;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/recipe/UniformBaseMultiBlockRecipe.java]:
```suggestion
        int minZ = getLastMatchingInDirection(target, location, Direction.BACKWARD.asVector3i()).z;
        int maxZ = getLastMatchingInDirection(target, location, Direction.FORWARD.asVector3i()).z;
```

File: src/main/java/org/terasology/multiBlock2/recipe/UniformBaseMultiBlockRecipe.java [hunk 13]
```diff
@@ -70,12 +74,12 @@ public T detectFormingMultiBlock(Vector3i location) {
...
-    protected abstract T createMultiBlockDefinition(Region3i multiBlockRegion);
+    protected abstract T createMultiBlockDefinition(BlockRegionc multiBlockRegion);
 
-    private Vector3i getLastMatchingInDirection(EntityRef targetEntity, Vector3i location, Vector3i direction) {
-        Vector3i result = location;
+    private Vector3i getLastMatchingInDirection(EntityRef targetEntity, Vector3ic location, Vector3ic direction) {
+        Vector3i result = new Vector3i(location);
         while (true) {
-            Vector3i testedLocation = new Vector3i(result.x + direction.x, result.y + direction.y, result.z + direction.z);
+            Vector3i testedLocation = result.add(direction, new Vector3i());
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/recipe/UniformBaseMultiBlockRecipe.java]: Same pattern with `tmp` vector as described above?

File: src/main/java/org/terasology/multiBlock2/recipe/UniformMultiBlockRecipe.java [hunk 14]
```diff
@@ -65,12 +68,12 @@ public T detectFormingMultiBlock(Vector3i location) {
...
-    protected abstract T createMultiBlockDefinition(Region3i multiBlockRegion);
+    protected abstract T createMultiBlockDefinition(BlockRegion multiBlockRegion);
 
-    private Vector3i getLastMatchingInDirection(Vector3i location, Vector3i direction) {
-        Vector3i result = location;
+    private Vector3i getLastMatchingInDirection(Vector3ic location, Vector3ic direction) {
+        Vector3i result = new Vector3i(location);
         while (true) {
-            Vector3i testedLocation = new Vector3i(result.x + direction.x, result.y + direction.y, result.z + direction.z);
+            Vector3i testedLocation = result.add(direction, new Vector3i());
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/recipe/UniformMultiBlockRecipe.java]: same pattern with `tmp` vector as described above?

File: src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java [hunk 15]
```diff
@@ -297,11 +299,10 @@ private void processLoadedMultiBlockMain(EntityRef mainBlockEntity, MultiBlockMa
...
 
-    private Region3i getChunkRegion(Vector3i chunkPos) {
-        return Region3i.createFromMinAndSize(
-                new Vector3i(chunkPos.x << ChunkConstants.POWER_X,
-                        chunkPos.y << ChunkConstants.POWER_Y,
-                        chunkPos.z << ChunkConstants.POWER_Z), ChunkConstants.CHUNK_SIZE);
+    private BlockRegion getChunkRegion(Vector3i chunkPos) {
+        return new BlockRegion(chunkPos.x << Chunks.POWER_X,
+                        chunkPos.y << Chunks.POWER_Y,
+                        chunkPos.z << Chunks.POWER_Z).setSize(Chunks.CHUNK_SIZE);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java]: Is there no utility in `Chunks` that does this? If there is not, should we add it? It's basically a lookup of the block region a specific chunks is spanning, right?

File: src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java (hunk 1 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java]: Yes, indeed. As we have a main location this simply becomes:

```java
return memberLocations.stream().reduce(new BlockRegion(mainLocation), BlockRegion::union, BlockRegion::union);
```

**Review by skaldarnar — CHANGES_REQUESTED** [review]

File: src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java (hunk 15 above)
Comment (skaldarnar) [reply src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java]: Please add a `//TODO: provide this as utility on Chunks?`

File: src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java [hunk 16]
```diff
@@ -290,18 +292,17 @@ private void processLoadedMultiBlockMain(EntityRef mainBlockEntity, MultiBlockMa
...
             multiBlockEntity.send(new MultiBlockLoaded(multiBlockMain.getMultiBlockType(), mainBlockEntity));
         }
     }
 
-    private Region3i getChunkRegion(Vector3i chunkPos) {
-        return Region3i.createFromMinAndSize(
-                new Vector3i(chunkPos.x << ChunkConstants.POWER_X,
-                        chunkPos.y << ChunkConstants.POWER_Y,
-                        chunkPos.z << ChunkConstants.POWER_Z), ChunkConstants.CHUNK_SIZE);
+    private BlockRegion getChunkRegion(Vector3i chunkPos) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/multiBlock2/system/MultiBlockServerSystem.java:302]: Please add a
```suggestion
    //TODO: provide this as utility on Chunks?
    private BlockRegion getChunkRegion(Vector3i chunkPos) {
```

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**